Proofs/Experiments/MemoryAliasing.lean

```diff
@@ -10,19 +10,107 @@ import Arm.Memory.MemoryProofs
+#guard_msgs in #print axioms subset_2
+
+/-- Show that smaller subsets are also subsets, even when moving base pointer. -/
+theorem subset_3 (l : mem_subset' a 16 b 16) : mem_subset' (a+6) 10 b 16 := by
+  simp_mem
+
+/-- info: 'MemSubset.subset_3' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms subset_3
+
+end MemSubset
+
+namespace MemSeparate
+
+/-- Reflexivity. -/
+theorem separate_1 (l : mem_separate' a 16 b 16) : mem_separate' a 16 b 16 := by
+  simp_mem
+
+/-- info: 'MemSeparate.separate_1' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms separate_1
+
+
+/-- Symmetry. -/
+theorem separate_2 (l : mem_separate' a 16 b 16) : mem_separate' b 16 a 16 := by
+  simp_mem
+
+/-- info: 'MemSeparate.separate_2' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms separate_2
+
+/-- Smaller subsets. -/
+theorem separate_3 (l : mem_separate' a 16 b 16) : mem_separate' b 10 a 10 := by
+  simp_mem
+
+/-- info: 'MemSeparate.separate_3' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms separate_3
+
+
+/-- sliding subset to the right. -/
+theorem separate_4 (l : mem_separate' a 16 b 16) (hab : a < b) :
+    mem_separate' a 17 (b+1) 15 := by
+  simp_mem
+
+/-- info: 'MemSeparate.separate_4' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms separate_4
+
+/-- shifts inside the arithmetic. -/
+theorem separate_5 {n : Nat} (hn : n ≠ 0)
+    (l : mem_separate' a (n <<< 4) b (n <<< 4))  :
+    mem_separate' a 16 b 16 := by
+  simp_mem
+
+/-- info: 'MemSeparate.separate_5' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms separate_5
+
+/-- shifts inside the arithmetic. -/
+theorem separate_6 {n : Nat} (hn : n ≠ 0)
+    (l : mem_separate' a (n <<< 4) b (n <<< 4))  :
+    mem_separate' a (n <<< 3 + 8) b (n <<< 4) := by
+  simp_mem
+
+/-- info: 'MemSeparate.separate_6' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms separate_6
+
+end MemSeparate
+
+
+theorem mem_automation_test_1
   (h_s0_src_dest_separate : mem_separate' src_addr  16 dest_addr 16) :
   read_mem_bytes 16 src_addr (write_mem_bytes 16 dest_addr blah s0) =
   read_mem_bytes 16 src_addr s0 := by
-  -- ⊢ read_mem_bytes 16 src_addr (write_mem_bytes 16 dest_addr blah s0) = read_mem_bytes 16 src_addr s0
+  simp only [memory_rules]
   simp_mem
-  -- ⊢ read_mem_bytes 16 src_addr s0 = read_mem_bytes 16 src_addr s0
   rfl
 
-/-- info: 'mem_automation_test' depends on axioms: [propext, Classical.choice, Quot.sound] -/
-#guard_msgs in #print axioms mem_automation_test
+/-- info: 'mem_automation_test_1' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms mem_automation_test_1
 
 theorem mem_automation_test_2
```

Comment:
One more test for a future PR: does the following simplify?
`read_mem_bytes n1 addr1 (write_mem_bytes n_ignore addr_ignore (write_mem_bytes n2 addr2 s))` where `[addr_ignore, n_ignore)` is separate from `[addr1, n1)` and `[addr1, n1)` is a subset of `[addr2, n2)`. 

Or something like that.

Reply:
Proved it anyway:

```lean
/-- reading from a region `[src_addr+1..10] ⊆ [src_addr..16]` with an
interleaved write `[ignore_addr..ignore_addr+ignore_n)`
-/
theorem mem_automation_test_3
  (h_no_wrap_src_region : mem_legal' src_addr 16)
  (h_s0_src_ignore_disjoint :
    mem_separate' src_addr  16
                  ignore_addr ignore_n) :
  read_mem_bytes 10 (src_addr + 1) (write_mem_bytes ignore_n ignore_addr blah s0) =
   read_mem_bytes 10 (src_addr + 1) s0 := by
  simp only [memory_rules]
  simp_mem
  rfl


/-- info: 'mem_automation_test_3' depends on axioms: [propext, Classical.choice, Quot.sound] -/
#guard_msgs in #print axioms mem_automation_test_3
```

Reply:
I meant something like this: note the two `simp_mem`s there and the `have`.

```
theorem mem_automation_test_4
  (h_no_wrap_src_region : mem_legal' src_addr 48)
  (h_s0_src_ignore_disjoint :
    mem_separate' src_addr  48
                  ignore_addr ignore_n) :
  read_mem_bytes 10 (1 + src_addr)
    (write_mem_bytes ignore_n ignore_addr blah 
      (write_mem_bytes 48 src_addr val s0)) =
   val.extractLsBytes 1 10 := by
  simp only [memory_rules]
  simp_mem
  simp_mem
  have : ((1 + src_addr).toNat - src_addr.toNat) = 1 := by
    unfold mem_legal' at h_no_wrap_src_region
    bv_omega
  simp only [this]
  done
```

Reply:
Hey, I used `simp_mem` for the first time here! :)

Reply:
Also re. `unfold mem_legal' at h_no_wrap_src_region` -- I think you have a lemma that gives an "omega" fact. Maybe I should've used that here.

Reply:
Added with TODO and golfed proof